src/algorithms/synthesis/dd_synthesis.cpp

```diff
@@ -52,4 +54,350 @@ namespace syrec {
+                            ctrl.insert(obj);
+                        } else {
+                            dd::Control obj;
+                            obj.qubit = static_cast<dd::Qubit>((i.size() + current->v) - j);
+                            obj.type  = dd::Control::Type::neg;
+                            ctrl.insert(obj);
+                        }
+                    }
+
+                    for (auto const& p2: p2SigVec) {
+                        for (std::size_t pj = 0; pj < p2.size(); pj++) {
+                            if (*(p2[pj])) {
+                                dd::Control obj;
+                                obj.qubit = static_cast<dd::Qubit>(current->v - (pj + 1));
+                                obj.type  = dd::Control::Type::pos;
+                                ctrl.insert(obj);
+                            } else {
+                                dd::Control obj;
+                                obj.qubit = static_cast<dd::Qubit>(current->v - (pj + 1));
+                                obj.type  = dd::Control::Type::neg;
+                                ctrl.insert(obj);
+                            }
+                        }
+
+                        auto q = qc::QuantumComputation((src.p->v) + 1U);
+                        q.x(current->v, ctrl);
+                        qc.x(current->v, ctrl);
+                        ctrl.clear();
+                        auto ddTest = dd::buildFunctionality(&q, dd);
+                        src         = dd->multiply(src, ddTest);
+                    }
+                }
+
+                synthesize(src, dd);
+
+            }
+
+            else {
+                unifyPath(src, current, p1SigVec, p2SigVec, indices, dd);
+            }
+        }
+    }
+
+    auto DDSynthesis::shiftingPaths(dd::mEdge& src, dd::mNode* current, std::unique_ptr<dd::Package<>>& dd) -> void {
+        auto srcCpy = src;
+
+        if (terminate(current)) {
+            return;
+        }
+
+        swapPaths(src, current, dd);
+
+        if (srcCpy == src) {
+            shiftUniquePaths(src, current, dd);
+        }
+    }
+
+    auto DDSynthesis::synthesize(dd::mEdge& src, std::unique_ptr<dd::Package<>>& dd) -> qc::QuantumComputation& {
+        std::vector<dd::mNode*> q;
+        std::vector<dd::mNode*> visited;
+
+        shiftingPaths(src, src.p, dd);
+
+        auto const* currNode = src.p;
+
+        while (src != dd->makeIdent(static_cast<dd::Qubit>((src.p->v) + 1U))) {
+            for (const auto& edge: currNode->e) {
+                if (!edge.isTerminal()) {
+                    auto it = std::find(visited.begin(), visited.end(), edge.p);
+
+                    if (it == visited.end()) {
+                        shiftingPaths(src, edge.p, dd);
+                        q.emplace_back(edge.p);
+                        visited.emplace_back(edge.p);
+                    }
+                }
+            }
+            currNode = q.front();
```

Comment:
It seems to me like you actually want to use a queue here, i.e. a FIFO data structure.

Reply:
done 👍🏻